lib/asar/fs-wrapper.ts

```diff
@@ -0,0 +1,779 @@
+import { Buffer } from 'buffer';
+import * as path from 'path';
+import * as util from 'util';
+
+const asar = process._linkedBinding('electron_common_asar');
+const v8Util = process._linkedBinding('electron_common_v8_util');
+
+const Module = require('module');
+
+const Promise: PromiseConstructor = global.Promise as any;
+
+const envNoAsar = process.env.ELECTRON_NO_ASAR &&
+    process.type !== 'browser' &&
+    process.type !== 'renderer';
+const isAsarDisabled = () => process.noAsar || envNoAsar;
+
+const internalBinding = (process as any).internalBinding;
+delete (process as any).internalBinding;
+
+const nextTick = (functionToCall: Function, args: any[] = []) => {
+  process.nextTick(() => functionToCall(...args));
+};
+
+// Cache asar archive objects.
+const cachedArchives = new Map();
+
+const getOrCreateArchive = (archivePath: string) => {
+  const isCached = cachedArchives.has(archivePath);
+  if (isCached) {
+    return cachedArchives.get(archivePath);
+  }
+
+  const newArchive = asar.createArchive(archivePath);
+  if (!newArchive) return null;
+
+  cachedArchives.set(archivePath, newArchive);
+  return newArchive;
+};
+
+// Separate asar package's path from full path.
+const splitPath = (archivePathOrBuffer: string | Buffer) => {
+  // Shortcut for disabled asar.
+  if (isAsarDisabled()) return { isAsar: false };
+
+  // Check for a bad argument type.
+  let archivePath = archivePathOrBuffer;
+  if (Buffer.isBuffer(archivePathOrBuffer)) {
+    archivePath = archivePathOrBuffer.toString();
+  }
+  if (typeof archivePath !== 'string') return { isAsar: false };
+
+  return asar.splitPath(path.normalize(archivePath));
+};
+
+// Convert asar archive's Stats object to fs's Stats object.
+let nextInode = 0;
+
+const uid = process.getuid != null ? process.getuid() : 0;
+const gid = process.getgid != null ? process.getgid() : 0;
+
+const fakeTime = new Date();
+const msec = (date: Date) => (date || fakeTime).getTime();
+
+const asarStatsToFsStats = function (stats: any) {
```

Comment:
we can probably type this, no? We know what the object's properties are: https://github.com/electron/electron/blob/602913cb4c98d102548e082c97401415849fe082/shell/common/api/electron_api_asar.cc#L71-L75